pkg/cmd/cli/cmd/tag.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	tagLong = `Tag existing images into image streams
+
+The tag command allows you to take an existing tag or image from an image
+stream, or a Docker image pull spec, and set it as the most recent image for a
+tag in 1 or more other image streams. It is similar to the 'docker tag'
+command, but it operates on image streams instead.`
+
+	tagExample = `  // Tag the current image for the image stream 'openshift/ruby' and tag '2.0' into the image stream 'yourproject/ruby with tag 'tip':
+  $ %[1]s tag openshift/ruby:2.0 yourproject/ruby:tip
+
+  // Tag a specific image:
+  $ %[1]s tag openshift/ruby@sha256:6b646fa6bf5e5e4c7fa41056c27910e679c03ebe7f93e361e6515a9da7e258cc yourproject/ruby:tip
+
+  // Tag an external Docker image:
+  $ %[1]s tag --source=docker openshift/origin:latest yourproject/ruby:tip`
+)
+
+// NewCmdTag implements the OpenShift cli tag command.
+func NewCmdTag(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	var sourceKind string
+
+	cmd := &cobra.Command{
+		Use:     "tag [--source=SOURCETYPE] SOURCE DEST [DEST ...]",
+		Short:   "Tag existing images into image streams",
+		Long:    tagLong,
+		Example: fmt.Sprintf(tagExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunTag(f, out, cmd, args, sourceKind)
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringVar(&sourceKind, "source", sourceKind, "Optional hint for the source type; valid values are 'imagestreamtag', 'istag', 'imagestreamimage', 'isimage', and 'docker'")
+
+	return cmd
+}
+
+func parseStreamName(name, defaultNamespace string) (string, string, error) {
+	if !strings.Contains(name, "/") {
+		return defaultNamespace, name, nil
+	}
+
+	parts := strings.Split(name, "/")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid image stream %q", name)
+	}
+
+	namespace := parts[0]
+	if len(namespace) == 0 {
+		return "", "", fmt.Errorf("invalid namespace %q for image stream %q", namespace, name)
+	}
+
+	streamName := parts[1]
+	if len(streamName) == 0 {
+		return "", "", fmt.Errorf("invalid name %q for image stream %q", streamName, name)
+	}
+
+	return namespace, streamName, nil
+}
+
+// RunTag contains all the necessary functionality for the OpenShift cli tag command.
+func RunTag(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string, sourceKind string) error {
+	if len(args) < 2 {
+		return cmdutil.UsageError(cmd, "you must specify a source and at least one destination")
+	}
+
+	// DockerImage isn't in RESTMapper
+	if len(sourceKind) > 0 {
+		switch strings.ToLower(sourceKind) {
+		case "docker", "dockerimage":
+			sourceKind = "DockerImage"
+		}
+	}
+
+	// try RESTMapper if it's not DockerImage
+	if len(sourceKind) > 0 && sourceKind != "DockerImage" {
+		original := sourceKind
+
+		mapper, _ := f.Object()
+		_, kind, _ := mapper.VersionAndKindForResource(sourceKind)
```

Comment:
Do this the other way around - just call mapper, if you get an error it's not recognized, then check docker image.